rmw_fastrtps_cpp/src/register_node.cpp

```diff
@@ -0,0 +1,217 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_fastrtps_cpp/register_node.hpp"
+
+#include <cassert>
+#include <memory>
+
+#include "rmw/error_handling.h"
+#include "rmw/init.h"
+#include "rmw/qos_profiles.h"
+
+#include "rmw_dds_common/context.hpp"
+#include "rmw_dds_common/msg/participant_entities_info.hpp"
+
+#include "rmw_fastrtps_cpp/identifier.hpp"
+#include "rmw_fastrtps_cpp/publisher.hpp"
+#include "rmw_fastrtps_cpp/subscription.hpp"
+
+#include "rmw_fastrtps_shared_cpp/custom_participant_info.hpp"
+#include "rmw_fastrtps_shared_cpp/participant.hpp"
+#include "rmw_fastrtps_shared_cpp/publisher.hpp"
+#include "rmw_fastrtps_shared_cpp/subscription.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_context_impl.hpp"
+
+#include "rosidl_typesupport_cpp/message_type_support.hpp"
+
+#include "listener_thread.hpp"
+
+using rmw_dds_common::msg::ParticipantEntitiesInfo;
+
+static
+rmw_ret_t
+init_context_impl(rmw_context_t * context)
+{
+  rmw_publisher_options_t publisher_options = rmw_get_default_publisher_options();
+  rmw_subscription_options_t subscription_options = rmw_get_default_subscription_options();
+
+  // This is currently not implemented in fastrtps
+  subscription_options.ignore_local_publications = true;
+
+  std::unique_ptr<rmw_dds_common::Context> common_context(
+    new(std::nothrow) rmw_dds_common::Context());
+  if (!common_context) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  std::unique_ptr<CustomParticipantInfo, std::function<void(CustomParticipantInfo *)>>
+  participant_info(
+    rmw_fastrtps_shared_cpp::create_participant(
+      eprosima_fastrtps_identifier,
+      context->options.domain_id,
+      &context->options.security_options,
+      (context->options.localhost_only == RMW_LOCALHOST_ONLY_ENABLED) ? 1 : 0,
+      common_context.get()),
+    [&](CustomParticipantInfo * participant_info) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_participant(participant_info)) {
+        fprintf(
+          stderr, "Failed to destroy participant after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!participant_info) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  rmw_qos_profile_t qos = rmw_qos_profile_default;
+
+  qos.avoid_ros_namespace_conventions = false;  // Change this to true after testing
```

Comment:
What is this about?

Reply:
Do we want the topic where the "ros discovery information" is published to be accessible by the user?
- Should it be a ROS hidden topic?
- Should it be a DDS topic name (no mangling applied, i.e. no `rt/...`)?
- Should it be just a normal ROS topic?

I'm inclined to option 1 or option 2, and I slightly prefer option 2, as not all the implementations will use this topic.

Reply:
We discussed this didn't we?

I would say no to the normal or hidden ROS topic, because if that were the case we should do it in `rcl` instead. We chose not to do that (if I remember correctly) because it would prevent middlewares from doing it in a more optimized way.

I think option 2 as well, but I think all the implementations should use this topic if they want to be cross-vendor compatible.

----

So my comment is more like, "well did we test it and are we going to flip it to true or not"?

I think we should flip it, seems like that would be option 2. But either way that comment should be removed, or changed. As-is it is written like a TODO.

Reply:
What happened here? Looks like nothing changed.

Reply:
> I think we should flip it, seems like that would be option 2. But either way that comment should be removed, or changed. As-is it is written like a TODO.

Flipped.